MTAConfigGenerator.py

```diff
@@ -1,152 +1,398 @@
+
+class MTAConfigGenerator():
+  allowed_ignore_list = {}
+
+  def __init__(self, policy_config, fixup=False, ignore_list=[]):
+    self.policy_config = policy_config
+    self.fixup = fixup
+    self.ignore_list = ignore_list
+
+    for err in self.ignore_list:
+      if err not in self.allowed_ignore_list:
+        raise ValueError("Unsupported error to be ignored: %s" % err)
+
+    self.changed_files = []
+
+    self.policy_defs_file = ""
+    self.policy_defs = ""
+
+  def build_defs(self):
+    """
+    Build the new STARTTLS enforcing policy definitions.
+    
+    Must be implemented in child class.
+    """
+
+    raise NotImplementedError()
+
+  def show_defs(self):
+    """Print the new policy definitions."""
+
+    sys.stdout.write(self.policy_defs)
+
+  def update_defs(self):
+    """
+    Update STARTTLS enforcing policy definitions only
+    without changing MTA's main configuration files.
+
+    Return:
+    - True if the new policy is immediately activated;
+    - False if an external action is required to instruct the
+      MTA to reload it.
+
+    Should be implemented in child class; the following
+    is only a basic implementation that save the new
+    policy definitions into the defs_file.
+    """
+    if self.policy_defs != "":
+      f = open(self.policy_defs_file, "w")
+      f.write(self.policy_defs)
+      f.close()
+
+    return False
+
+  def build_general_config(self):
+    """
+    Build changes needed to adapt MTA's main configuration
+    to what's needed by STARTTLS-Everywhere.
+    
+    Return True if changes are needed, otherwise False.
+
+    Files that need to be modified must be added to self.changed_files.
+    
+    Must be implemented in child class.
+    """
+
+    raise NotImplementedError()
+
+  def show_new_general_config(self):
+    """
+    Print the new configuration for each file that needs
+    to be changed.
+    """
+
+    for F in self.changed_files:
+      print("File: {}".format(F.path))
+      sys.stdout.write(F.new_data)
+
+  def fix_general_config(self):
+    """
+    Update only MTA's main configuration files, adapting them
+    to what's needed by STARTTLS-Everywhere.
+
+    Return:
+    - True if the new configuration is immediately activated;
+    - False if an external action is required to instruct the
+      MTA to reload it.
+    """
+    for F in self.changed_files:
+      F.save()
+
+    return False
+
+  def show_new_general_config_diff(self):
+    """
+    Show a diff between current and new MTA's general config.
+
+    It uses the command line provided in STARTTLS-Everywhere
+    configuration.
+    """
+
+    diff_tpl = Config.get("general","diff_cmd")
+
+    if diff_tpl.strip() == "":
+      raise ValueError("The diff external command is missing; "
+                       "please set the diff_cmd parameter in the general "
+                       "STARTTLS-Everywhere configuration.")
+
+    for F in self.changed_files:
```

Comment:
we should probably consider adding something like this for verbose operation, and backup changed files in the directory with an iso datetime extension or something.